batch2/batch/driver/scheduler.py

```diff
@@ -19,6 +20,59 @@ def __init__(self, app):
         asyncio.ensure_future(self.loop('cancel_loop', self.cancel_state_changed, self.cancel_1))
         asyncio.ensure_future(self.bump_loop())
 
+    async def compute_fair_share(self):
+        free_cores_mcpu = sum([worker.free_cores_mcpu for worker in self.inst_pool.healthy_instances_by_free_cores])
+
+        sorted_running_cores_mcpu = sortedcontainers.SortedSet(key=lambda x: x[1])
+        sorted_total_cores_mcpu = sortedcontainers.SortedSet(key=lambda x: x[1])
+
+        running_cores_mcpu = {}
+        allocated_cores = {}
+
+        records = self.db.execute_and_fetchall(
+            '''
+SELECT user, ready_cores_mcpu, running_cores_mcpu
+FROM user_resources
+WHERE ready_cores_mcpu > 0;
+''')
+
+        async for record in records:
+            running_cores_mcpu[record['user']] = record['running_cores_mcpu']
+            sorted_running_cores_mcpu.add(
+                (record['user'], record['running_cores_mcpu']))
+            sorted_total_cores_mcpu.add(
+                (record['user'], record['running_cores_mcpu'] + record['ready_cores_mcpu']))
+
+        target = 0
+        allocating_users = set()
+        while sorted_total_cores_mcpu:
+            lowest_running_user, lowest_running = sorted_running_cores_mcpu[0]
+            lowest_total_user, lowest_total = sorted_total_cores_mcpu[0]
+
+            allocation = min(lowest_running, lowest_total)
+
+            n_allocating_users = len(allocating_users)
+            cores_to_allocate = n_allocating_users * (allocation - target)
+
+            if cores_to_allocate > free_cores_mcpu:
+                target += free_cores_mcpu / n_allocating_users
+                break
+
+            target = allocation
+            free_cores_mcpu -= cores_to_allocate
+
+            if lowest_running >= lowest_total:
```

Comment:
I would do this in a loop before allocating.  In particular, you might have to retire several allocating users or move several users from pending to allocating before making an actual allocation.  So I imagined:

```
  if (lowest total at mark):
    remove lowest total user from allocating
  elif (mark at lowest running):
    move lowest running from pending to allocating
  else:
    allocate
```